### PR TITLE
Update UID2-Shared to Support KeysetKey Encryption/Decryption

### PR DESCRIPTION
**Pull request by zaiweidu**
1. Added encrypt/decrypt wrappers in UID2-Shared to support KeysetKey class in UID2-Operator.
2. Fixed broken unit tests and other minor typos.

### CODE REVIEW

File: src/main/java/com/uid2/shared/encryption/AesCbc.java [hunk 1]
```diff
@@ -35,6 +36,28 @@ public static EncryptedPayload encrypt(String s, EncryptionKey key) {
         }
     }
 
+    public static EncryptedPayload encrypt(byte[] b, KeysetKey key) {
+        try {
+            final SecretKey k = new SecretKeySpec(key.getKeyBytes(), "AES");
```
Comment (jon8787) [inline src/main/java/com/uid2/shared/encryption/AesCbc.java:41]: is this code copied from somewhere else? Can we make reusable functions instead?

File: src/main/java/com/uid2/shared/encryption/AesCbc.java [hunk 2]
```diff
@@ -44,6 +67,15 @@ public static byte[] decrypt(byte[] encryptedBytes, EncryptionKey key) {
         }
     }
 
+    public static byte[] decrypt(byte[] encryptedBytes, KeysetKey key) {
```
Comment (jon8787) [inline src/main/java/com/uid2/shared/encryption/AesCbc.java:70]: with these and below, can we make reusable functions? eg pass key.GetKeyBytes() into a common function that is used by new KeysetKeys and old keys

File: src/main/java/com/uid2/shared/encryption/AesCbc.java (hunk 1 above)
Comment (jon8787) [inline src/main/java/com/uid2/shared/encryption/AesCbc.java:41]: can we make a common function so we don't duplicate this logic?

File: src/main/java/com/uid2/shared/encryption/AesGcm.java [hunk 3]
```diff
@@ -32,6 +33,23 @@ public static EncryptedPayload encrypt(String s, EncryptionKey key) {
         }
     }
 
+    public static EncryptedPayload encrypt(byte[] b, KeysetKey key) {
+        try {
+            byte[] encrypted = encrypt(b, key.getKeyBytes());
```
Comment (jon8787) [inline src/main/java/com/uid2/shared/encryption/AesGcm.java:38]: Can we make common functions or remove the old functions?

File: src/main/java/com/uid2/shared/store/IKeysetSnapshot.java [hunk 4]
```diff
@@ -1,8 +1,11 @@
 package com.uid2.shared.store;
 
 import com.uid2.shared.auth.ClientKey;
+import com.uid2.shared.auth.Keyset;
 import com.uid2.shared.model.KeysetKey;
+import com.uid2.shared.store.ACLMode.MissingAclMode;
 
 public interface IKeysetSnapshot {
-    boolean canClientAccessKey(ClientKey clientKey, KeysetKey key);
+    boolean canClientAccessKey(ClientKey clientKey, KeysetKey key, MissingAclMode accessMethod);
```
Comment (jon8787) [inline src/main/java/com/uid2/shared/store/IKeysetSnapshot.java:9]: can we remove this file? It seems to have the same interface as KeysetSnapshot

File: src/main/java/com/uid2/shared/store/IKeysetSnapshot.java [hunk 5]
```diff
@@ -1,8 +1,15 @@
 package com.uid2.shared.store;
 
 import com.uid2.shared.auth.ClientKey;
+import com.uid2.shared.auth.Keyset;
 import com.uid2.shared.model.KeysetKey;
+import com.uid2.shared.store.ACLMode.MissingAclMode;
+
+import java.util.Map;
 
 public interface IKeysetSnapshot {
```
Comment (jon8787) [inline src/main/java/com/uid2/shared/store/IKeysetSnapshot.java:10]: let's remove this file if it's just a duplication of KeysetSnapshot's interface

File: src/main/java/com/uid2/shared/auth/Keyset.java [hunk 6]
```diff
@@ -61,7 +61,7 @@ public boolean isDefault() {
 
     public boolean canBeAccessedBySite(Integer siteId) {
         if(!isEnabled) return false;
-        if(allowedSites == null) return false;
+        if(allowedSites == null || allowedSites.isEmpty()) return false;
```
Comment (jon8787) [inline src/main/java/com/uid2/shared/auth/Keyset.java:64]: do we have a test case that failed when this wasn't there?

File: src/main/java/com/uid2/shared/auth/KeysetSnapshot.java [hunk 7]
```diff
@@ -20,5 +21,18 @@ public boolean canClientAccessKey(ClientKey clientKey, KeysetKey key) {
         return keyset.canBeAccessedBySite(clientKey.getSiteId());
     }
 
-    public Map<Integer, Keyset> getAllKeysets() { return keysets; }
+    @Override
+    public boolean canClientAccessKey(ClientKey clientKey, KeysetKey key, MissingAclMode accessMethod) {
+        Keyset keyset = keysets.get(key.getKeysetId());
```
Comment (jon8787) [inline src/main/java/com/uid2/shared/auth/KeysetSnapshot.java:26]: 1. Why do we need to keep the old "canClientAccessKey(ClientKey clientKey, KeysetKey key)" method?
2. Assuming we do need it, I think this method should have all the functionality, and canClientAccessKey(ClientKey clientKey, KeysetKey key) should simply call canClientAccessKey(ClientKey clientKey, KeysetKey key, MissingAclMode accessMethod) with the appropriate default.
3. Do we have test coverage for both these methods?

File: src/main/java/com/uid2/shared/store/IKeysetProvider.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package com.uid2.shared.store;
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/store/IKeysetProvider.java:1]: Why did you add this back in?

File: src/main/java/com/uid2/shared/encryption/AesGcm.java (hunk 3 above)
Comment (cody-constine-ttd) [reply src/main/java/com/uid2/shared/encryption/AesGcm.java:20]: Did you test that the old functions are safe to remove? This is technically a breaking changed to shared, so we should do a major release

File: src/main/java/com/uid2/shared/store/KeysetKeyStoreSnapshot.java [hunk 9]
```diff
@@ -30,7 +30,7 @@ public List<KeysetKey> getActiveKeysetKeys() {
     public KeysetKey getActiveKey(int keysetId, Instant now) {
         List<KeysetKey> keysetKeys = keysetMap.get(keysetId);
         if(keysetKeys == null || keysetKeys.isEmpty()) return null;
-        int keysetKeysIndex = Utils.upperBound(keysetKeys, now, (ts, k) -> ts.isBefore(k.getActivates()));
+        int keysetKeysIndex = Utils.upperBound(keysetKeys, now, (ts, k) -> !ts.isAfter(k.getActivates())); // before || equal
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/store/KeysetKeyStoreSnapshot.java:33]: Why did you make this change? Can you explain the difference in logic here? Looks to be the same logic just more complex

File: src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java [hunk 10]
```diff
@@ -144,11 +143,11 @@ public void loadContentMultipleEntries() throws Exception {
         Assert.assertFalse(canAccessKey(2, 3));
         //Only 4 can access null list
         Assert.assertTrue(canAccessKey(4,4));
-        Assert.assertFalse(canAccessKey(2, 4));
+        Assert.assertTrue(canAccessKey(2, 4));
```
Comment (cody-constine-ttd) [inline src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java:146]: We should be doing test with both ALLOW_ALL and DENY_ALL modes here

File: src/main/java/com/uid2/shared/auth/Keyset.java [hunk 11]
```diff
@@ -79,7 +78,7 @@ public boolean equals(Object o) {
                 && this.isEnabled == b.isEnabled
                 && this.isDefault == b.isDefault;
 
-        if(this.allowedSites == null || this.allowedSites == null) {
+        if(this.allowedSites == null || b.allowedSites == null) {
```
Comment (jon8787) [inline src/main/java/com/uid2/shared/auth/Keyset.java:81]: A bit scary that none of us saw the original bug. You can use intellij to generate these methods.  Maybe we should have test coverage of this method?

File: src/main/java/com/uid2/shared/store/KeysetKeyStoreSnapshot.java (hunk 9 above)
Comment (jon8787) [reply src/main/java/com/uid2/shared/store/KeysetKeyStoreSnapshot.java:33]: also if this change is intentional, let's add a test case

File: src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java [hunk 12]
```diff
@@ -139,16 +138,24 @@ public void loadContentMultipleEntries() throws Exception {
...
-        Assert.assertFalse(canAccessKey(2, 4));
+        Assert.assertTrue(canAccessKey(4,4, MissingAclMode.ALLOW_ALL));
+        Assert.assertTrue(canAccessKey(4,4, MissingAclMode.DENY_ALL));
+        Assert.assertTrue(canAccessKey(2, 4, MissingAclMode.ALLOW_ALL));
+        Assert.assertFalse(canAccessKey(2, 4, MissingAclMode.DENY_ALL));
         //Can still access if there is a duplicate
-        Assert.assertTrue(canAccessKey(1, 5));
+        Assert.assertTrue(canAccessKey(1, 5, MissingAclMode.ALLOW_ALL));
+        Assert.assertTrue(canAccessKey(1, 5, MissingAclMode.DENY_ALL));
         // Only 6 can access its empty list
```
Comment (jon8787) [inline src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java:155]: this comment is out of date

File: src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java [hunk 13]
```diff
@@ -139,16 +138,24 @@ public void loadContentMultipleEntries() throws Exception {
...
         //Only 4 can access null list
-        Assert.assertTrue(canAccessKey(4,4));
-        Assert.assertFalse(canAccessKey(2, 4));
+        Assert.assertTrue(canAccessKey(4,4, MissingAclMode.ALLOW_ALL));
+        Assert.assertTrue(canAccessKey(4,4, MissingAclMode.DENY_ALL));
+        Assert.assertTrue(canAccessKey(2, 4, MissingAclMode.ALLOW_ALL));
+        Assert.assertFalse(canAccessKey(2, 4, MissingAclMode.DENY_ALL));
         //Can still access if there is a duplicate
-        Assert.assertTrue(canAccessKey(1, 5));
+        Assert.assertTrue(canAccessKey(1, 5, MissingAclMode.ALLOW_ALL));
```
Comment (jon8787) [inline src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java:153]: should also ensure duplicates don't mess with denied sites. eg test for "canAccessKey(4, 5)"

File: src/main/java/com/uid2/shared/encryption/AesCbc.java (hunk 1 above)
Comment (zaiweidu) [reply src/main/java/com/uid2/shared/encryption/AesCbc.java:18]: test

File: src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java [hunk 14]
```diff
@@ -139,16 +138,34 @@ public void loadContentMultipleEntries() throws Exception {
...
+        Assert.assertTrue(canAccessKey(2, 5, MissingAclMode.ALLOW_ALL));
+        Assert.assertTrue(canAccessKey(2, 5, MissingAclMode.DENY_ALL));
+        Assert.assertFalse(canAccessKey(4, 5, MissingAclMode.ALLOW_ALL));
+        Assert.assertFalse(canAccessKey(4, 5, MissingAclMode.DENY_ALL));
+        Assert.assertTrue(canAccessKey(5, 5, MissingAclMode.ALLOW_ALL));
+        Assert.assertTrue(canAccessKey(5, 5, MissingAclMode.DENY_ALL));
+        // empty list
+        Assert.assertTrue(canAccessKey(6,6, MissingAclMode.ALLOW_ALL));
+        Assert.assertTrue(canAccessKey(6,6, MissingAclMode.DENY_ALL));
+        Assert.assertTrue(canAccessKey(5,6, MissingAclMode.ALLOW_ALL));
```
Comment (jon8787) [inline src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java:166]: I think this isn't correct - if a list is empty, it means we don't want to share with anyone, including ID_READER

File: src/main/java/com/uid2/shared/auth/Keyset.java (hunk 11 above)
Comment (zaiweidu) [reply src/main/java/com/uid2/shared/auth/Keyset.java:81]: It was hard to found.

File: src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java (hunk 14 above)
Comment (zaiweidu) [reply src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java:166]: @jon8787 @cody-constine-ttd - The new logic here is:
(1) if a keyset’s allowed_list is null or empty, allow or deny access will be controlled by MissingAclMode enum.
(2) keys belong to the site_id of the keyset, will automatically get access.

File: src/main/java/com/uid2/shared/auth/Keyset.java (hunk 11 above)
Comment (jon8787) [reply src/main/java/com/uid2/shared/auth/Keyset.java:81]: how did you find it?

**Review by zaiweidu — COMMENTED** [review]
push a new commit to address comments

**Review by jon8787 — APPROVED** [review]
